### PR TITLE
Update client to include new project params and override options

### PR DESCRIPTION
**Pull request by drew**
This pull request updates the `Project` class to support `detection_mode`, `entity_stream` and custom query and header params.

Custom headers and params are captured via `kwargs`. On the `Client`, I tried to push down param and header extraction as far down as possible into the request chain.

`_iter_records` is a notable exception. There were enough params in the function I figured it made sense to leave `entity_stream` a top-level parameter.

### CODE REVIEW

**Comment by drew** [comment]
- [ ] update `Project.iter_records()` to include `entity_stream` parameter.

File: src/gretel_client/client.py [hunk 1]
```diff
@@ -259,6 +295,7 @@ def _write_records(
             t = threading.Thread(
                 target=self._write_record_batch,
                 args=(project, write_queue, error_list, thread_event),
+                kwargs=kwargs,
```
Comment (johntmyers) [inline src/gretel_client/client.py:298]: does this need to be `**kwargs`?

File: src/gretel_client/client.py [hunk 2]
```diff
@@ -345,15 +391,22 @@ def __iter_records(
         start_key = "oldest" if forward else "newest"
         records_seen = 0
         last = position
+
+        # setup base params for streams request
+        headers, params = self._headers_params_from_kwargs(**kwargs)
+        params.update({"with_meta": "yes", "flatten": "yes"})
+        if entity_stream:
```
Comment (johntmyers) [inline src/gretel_client/client.py:398]: nit: can we do a `if entity_stream is not None` here, in the off chance an empty string is provided, etc.

File: src/gretel_client/client.py [hunk 3]
```diff
@@ -323,20 +360,29 @@ def noop(x):
...
-            self._get(f"streams/records/{project}", params).get(DATA).get(RECORDS, [])
+            self._get(f"streams/records/{project}", **kwargs)
+            .get(DATA)
+            .get(RECORDS, [])
         )
 
     def __iter_records(
         self,
         project: str,
+        entity_stream: str = None,
```
Comment (johntmyers) [inline src/gretel_client/client.py:380]: `Optional[str]` just to be consistent w/ the neighboring param hints

File: src/gretel_client/client.py [hunk 4]
```diff
@@ -323,20 +360,29 @@ def noop(x):
...
         position: Optional[str] = None,
         post_process: Callable = None,
```
Comment (johntmyers) [inline src/gretel_client/client.py:382]: `Optional` here too even though it wasn't part of the PR

File: src/gretel_client/client.py [hunk 5]
```diff
@@ -397,11 +450,13 @@ def _iter_records(
         self,
         *,
         project: str,
+        entity_stream: str = None,
         position: Optional[str] = None,
         post_process: Callable = None,
```
Comment (johntmyers) [inline src/gretel_client/client.py:455]: `Optional` up these ones too

File: src/gretel_client/projects.py [hunk 6]
```diff
@@ -132,34 +157,69 @@ def send(self, data: Union[List[dict], dict]) -> Tuple[list, list]:
...
         """
-        ret = self.client._write_record_sync(self.name, data)["data"]
+        kwargs = self._patch_params(kwargs, "detection_mode", detection_mode)
+        ret = self.client._write_record_sync(self.name, data, **kwargs)["data"]
         return ret["success"], ret["failure"]
 
-    def send_bulk(self, data: Union[List[dict], dict]) -> "WriteSummary":
+    def _patch_params(self, kwargs: dict, key: str, value: str) -> dict:
+        """Patch in named arguments into query param **kwarg type dictionary"""
+        assert isinstance(kwargs, dict)
```
Comment (johntmyers) [inline src/gretel_client/projects.py:177]: Don't use an assertion outside of a test here, we should check for this not being a `dict` then raise a `ValueError`

File: src/gretel_client/client.py (hunk 1 above)
Comment (drew) [reply src/gretel_client/client.py:298]: I believe so. `Thread.args` and `Thread.kwargs` represents `*args` and `**kwargs` when constructing the call to `target`. I was losing the named arguments when passing `**kwards` into `args`.

File: src/gretel_client/client.py (hunk 2 above)
Comment (drew) [reply src/gretel_client/client.py]: Updated.

File: src/gretel_client/client.py (hunk 3 above)
Comment (drew) [reply src/gretel_client/client.py]: ✅ fixed

File: src/gretel_client/client.py (hunk 4 above)
Comment (drew) [reply src/gretel_client/client.py:382]: updated.

File: src/gretel_client/client.py (hunk 5 above)
Comment (drew) [reply src/gretel_client/client.py]: 👍

File: src/gretel_client/projects.py (hunk 6 above)
Comment (drew) [reply src/gretel_client/projects.py]: good catch. fixed.

**Pull request merged**